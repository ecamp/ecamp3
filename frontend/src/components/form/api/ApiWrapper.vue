<!--
Wrapper component for form components to save data back to API
-->

<template>
  <v-form
    :class="['d-flex','flex-wrap',{'api-wrapper--inline':!autoSave && !readonly && !separateButtons, 'my-4':!noMargin}]"
    @submit.prevent="onEnter">
    <slot
      :localValue="localValue"
      :label="label"
      :hasServerError="hasServerError"
      :hasLoadingError="hasLoadingError"
      :hasValidationError="hasValidationError"
      :errorMessages="errorMessages"
      :isSaving="isSaving"
      :isLoading="isLoading"
      :autoSave="autoSave"
      :readonly="readonly || !hasFinishedLoading"
      :status="status"
      :on="eventHandlers" />

    <div
      v-if="!autoSave && !readonly"
      :class="['d-flex', {'my-1 ml-auto': separateButtons}]">
      <v-btn
        :disabled="disabled || !hasFinishedLoading"
        small
        elevation="0"
        :class="{'mr-1': separateButtons}"
        :tile="!separateButtons"
        :height="separateButtons ? '' : 'auto'"
        @click="reset">
        Reset
      </v-btn>

      <v-btn
        type="submit"
        :color="hasServerError ? 'error' : 'primary'"
        small
        elevation="0"
        :disabled="disabled || !hasFinishedLoading || hasValidationError"
        class="v-btn--last-instance"
        :height="separateButtons ? '' : 'auto'"
        :loading="isSaving"
        @click="save">
        {{ hasServerError ? 'Retry' : 'Save' }}
      </v-btn>
    </div>
  </v-form>
</template>

<script>

import { debounce } from 'lodash'
import { apiPropsMixin } from '@/mixins/apiPropsMixin'
import { validate } from 'vee-validate'

export default {
  name: 'ApiWrapper',
  mixins: [apiPropsMixin],
  props: {
    separateButtons: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      localValue: null,
      isSaving: false,
      isLoading: false,
      showIconSuccess: false,
      dirty: false,
      hasValidationError: false,
      hasServerError: false,
      hasLoadingError: false,
      serverErrorMessage: '',
      loadingErrorMessage: '',
      validationErrorMessages: [],
      eventHandlers: {
        save: this.save,
        reset: this.reset,
        reload: this.reload,
        input: this.onInput
      }
    }
  },
  computed: {
    hasFinishedLoading () {
      return !this.isLoading && !this.hasLoadingError
    },
    errorMessages () {
      const errors = []

      // 1st priority: Loading error
      if (this.hasLoadingError) errors.push(this.loadingErrorMessage)

      // 2nd priority: Frontent validation
      if (this.hasValidationError) this.validationErrorMessages.forEach(err => errors.push(err))

      // 3rd priority: Server error (not displayed in case of frontend validation error)
      if (this.hasServerError) errors.push(this.serverErrorMessage)

      return errors
    },
    status: function () {
      if (this.isSaving) {
        return 'saving'
      } else if (this.showIconSuccess) {
        return 'success'
      } else {
        return 'init'
      }
    },
    debouncedSave () {
      return debounce(this.save, this.autoSaveDelay)
    },
    apiValue () {
      // return value from props if set explicitly
      if (this.value) {
        return this.value

      // avoid infinite reloading if loading from API has failed
      } else if (this.hasLoadingError) {
        return null

      // return value from API unless `value` is set explicitly
      } else {
        return this.api.get(this.uri)[this.fieldname]
      }
    }
  },
  watch: {
    apiValue: function (newValue, oldValue) {
      // override local value if it wasn't dirty
      if (!this.dirty || this.overrideDirty) {
        this.localValue = newValue
      }

      // clear dirty if outside value changes to same as local value (e.g. after save operation)
      if (this.localValue === newValue) {
        this.dirty = false
      }
    }
  },
  created () {
    // initial data load from API
    if (!this.value) this.reload()

    this.localValue = this.apiValue
  },
  methods: {
    async onInput (newValue) {
      this.localValue = newValue
      this.dirty = true

      if (this.required || this.validation) {
        const rules = []
        if (this.required) rules.push('required')
        if (this.validation) rules.push(this.validation)

        const result = await validate(newValue, rules.join('|'), {
          name: this.label
        })

        this.hasValidationError = !result.valid
        this.validationErrorMessages = result.errors
      }

      if (this.autoSave) {
        this.debouncedSave()
      }
    },
    // reload data from API (doesn't force loading from server if available locally)
    reload () {
      this.isLoading = true
      this.resetErrors()

      // initial data load from API
      this.api.get(this.uri)._meta.load.then(() => {
        this.isLoading = false
      }).catch(error => {
        this.isLoading = false
        this.hasLoadingError = true
        this.loadingErrorMessage = error.message
      })
    },
    reset () {
      this.localValue = this.apiValue
      this.resetErrors()
    },
    resetErrors () {
      this.dirty = false
      this.hasLoadingError = false
      this.hasValidationError = false
      this.hasServerError = false
      this.serverErrorMessage = null
    },
    onEnter () {
      if (!this.autoSave) {
        this.save()
      }
    },
    save () {
      // abort saving if component is in readonly or disabled state
      // this is here for safety reasons, should not be triggered if the wrapped component behaves normally
      if (this.readonly || this.disabled) {
        return
      }

      // abort saving in case of validation errors
      if (this.hasValidationError) {
        return
      }

      // reset all dirty flags and start saving
      this.resetErrors()
      this.isSaving = true

      this.api.patch(this.uri, { [this.fieldname]: this.localValue }).then(() => {
        this.isSaving = false
        this.showIconSuccess = true
        setTimeout(() => { this.showIconSuccess = false }, 2000)
      }, (error) => {
        this.isSaving = false

        // 422 validation error
        if (error.name === 'ServerException' && error.response && error.response.status === 422) {
          this.serverErrorMessage = 'Validation error: '
          const validationMessages = error.response.data.validation_messages[this.fieldname]
          Object.keys(validationMessages).forEach((key) => {
            this.serverErrorMessage = this.serverErrorMessage + validationMessages[key] + '. '
          })
        } else {
          this.serverErrorMessage = error.message
        }

        this.hasServerError = true
      })
    }
  }

}
</script>

<style lang="scss" scoped>
  .api-wrapper--inline .v-btn--last-instance {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
  .api-wrapper--inline .v-btn {
    border-top: 1px solid rgba(0, 0, 0, 0.38);
    border-bottom: 1px solid rgba(0, 0, 0, 0.38);
  }
</style>

<style lang="scss" >
  .api-wrapper--inline .v-text-field {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
</style>
