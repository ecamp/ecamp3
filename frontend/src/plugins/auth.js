import axios from 'axios'
import { apiStore } from '@/plugins/store'
import router from '@/router'
import Cookies from 'js-cookie'

axios.interceptors.response.use(null, error => {
  if (error.status === 401) {
    logout().then(() => {})
  }
  return Promise.reject(error)
})

function getJWTPayloadFromCookie () {
  const jwtHeaderAndPayload = Cookies.get('jwt_hp')
  if (!jwtHeaderAndPayload) return ''

  return jwtHeaderAndPayload.split('.')[1]
}

function parseJWTPayload (payload) {
  if (!payload) return {}
  const base64 = payload.replace(/-/g, '+').replace(/_/g, '/')
  const jsonPayload = decodeURIComponent(atob(base64).split('').map(function (c) {
    return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2)
  }).join(''))

  return JSON.parse(jsonPayload)
}

function getJWTExpirationTimestamp () {
  return (parseJWTPayload(getJWTPayloadFromCookie()).exp ?? 0) * 1000
}

export function isLoggedIn () {
  return Date.now() < getJWTExpirationTimestamp()
}

async function login (username, password) {
  const url = await apiStore.href(apiStore.get(), 'login')
  return apiStore.post(url, { username: username, password: password }).then(() => {
    return isLoggedIn()
  })
}

async function resetPasswordRequest (email, token) {
  const url = await apiStore.href(apiStore.get(), 'resetPassword')
  return apiStore.post(url, { email: email, token: token })
}

async function resetPassword (id, password, token) {
  const url = await apiStore.href(apiStore.get(), 'resetPassword', { id: id })
  return apiStore.patch(url, { password: password, token: token })
}

function user () {
  if (!getJWTPayloadFromCookie()) {
    return null
  }
  const user = apiStore.get(parseJWTPayload(getJWTPayloadFromCookie()).user)
  user._meta.load = user._meta.load.catch(e => {
    if (e.response && [401, 403, 404].includes(e.response.status)) {
      // 401 means no complete token was submitted, so we may be missing the JWT signature cookie
      // 403 means we can theoretically interact in some way with the user, but apparently not read it
      // 404 means the user doesn't exist or we don't have access to it
      // Either way, we aren't allowed to access the user from the token, so it's best to ask the user
      // to log in again.
      auth.logout()
      return
    }
    throw e
  })
  return user
}

async function register (data) {
  const url = await apiStore.href(apiStore.get(), 'users')
  return apiStore.post(url, data)
}

async function redirectToOAuthLogin (provider) {
  let returnUrl = window.location.origin + router.resolve({ name: 'loginCallback' }).href

  const params = new URLSearchParams(window.location.search)
  if (params.has('redirect')) {
    returnUrl += '?redirect=' + params.get('redirect')
  }

  return apiStore.href(apiStore.get(), provider, { callback: encodeURI(returnUrl) }).then(url => {
    window.location.href = window.environment.API_ROOT_URL + url
  })
}

async function loginGoogle () {
  return redirectToOAuthLogin('oauthGoogle')
}

async function loginPbsMiData () {
  return redirectToOAuthLogin('oauthPbsmidata')
}

async function loginCeviDB () {
  return redirectToOAuthLogin('oauthCevidb')
}

export async function logout () {
  Cookies.remove('jwt_hp', { domain: window.environment.SHARED_COOKIE_DOMAIN })
  return router.push({ name: 'login' })
    .then(() => apiStore.purgeAll())
    .then(() => isLoggedIn())
}

export const auth = {
  isLoggedIn,
  login,
  register,
  loginGoogle,
  loginPbsMiData,
  loginCeviDB,
  logout,
  user,
  resetPasswordRequest,
  resetPassword
}

class AuthPlugin {
  install (Vue) {
    Object.defineProperties(Vue.prototype, {
      $auth: {
        get () {
          return auth
        }
      }
    })
  }
}

export default new AuthPlugin()
