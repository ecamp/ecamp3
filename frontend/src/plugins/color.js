import { ColorSpace, sRGB } from 'colorjs.io/fn'

export default {
  install: () => {
    ColorSpace.register(sRGB)
  },
}
