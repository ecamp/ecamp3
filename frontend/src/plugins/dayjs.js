import dayjs from '@/../../common/helpers/dayjs.js'

export default {
  install: (Vue) => {
    Object.defineProperties(Vue.prototype, {
      $date: {
        get () {
          return dayjs
        }
      }
    })
    Vue.dayjs = dayjs
  }
}
