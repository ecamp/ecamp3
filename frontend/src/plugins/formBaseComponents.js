// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
