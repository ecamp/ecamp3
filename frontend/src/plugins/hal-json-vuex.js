import HalJsonVuex from 'hal-json-vuex'
import axios from 'axios'
import { createStore } from 'vuex'
import { getEnv } from '@/environment.js'

export default {
  install: (app, options) => {
    // create store
    const store = createStore({
      state() {
        return {}
      },
    })
    app.use(store)

    const axiosInstance = axios.create({
      withCredentials: true,
      baseURL: getEnv().API_ROOT_URL,
      headers: { common: { Accept: 'application/hal+json' } },
    })
    axiosInstance.interceptors.request.use(function (config) {
      if (config.method === 'patch') {
        config.headers['Content-Type'] = 'application/merge-patch+json'
      }
      return config
    })

    // create and inject API
    const api = new HalJsonVuex(store, axiosInstance, {
      forceRequestedSelfLink: true,
    })

    app.use(api)
  },
}
