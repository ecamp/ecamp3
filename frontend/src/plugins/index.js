export { default as vuetify } from './vuetify'
export { default as store } from './store'
export { default as auth } from './auth'
export { default as filterLoading } from './filterLoading'
export { default as formBaseComponents } from './formBaseComponents'
export { default as ignoreNativeBindingWarnMessages } from './ignoreNativeBindingWarnMessages'
export { default as veeValidate } from './veeValidate'
