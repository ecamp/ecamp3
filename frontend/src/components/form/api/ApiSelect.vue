<!--
Displays a field as a e-select + write access via API wrapper
-->

<template>
  <api-wrapper
    v-slot="wrapper"
    v-bind="$props">
    <e-select
      :value="wrapper.localValue"
      :label="wrapper.label"
      v-bind="$attrs"
      :readonly="wrapper.readonly"
      :disabled="disabled"
      :error-messages="wrapper.errorMessages"
      :loading="wrapper.isSaving || wrapper.isLoading ? 'secondary' : false"
      outlined
      no-margin
      :filled="false"
      @input="wrapper.on.input">
      <template #append>
        <v-icon>mdi-menu-down</v-icon>
        <api-wrapper-append :wrapper="wrapper" />
      </template>
    </e-select>
  </api-wrapper>
</template>

<script>
import { apiPropsMixin } from '@/mixins/apiPropsMixin'
import ApiWrapper from './ApiWrapper'
import ApiWrapperAppend from './ApiWrapperAppend'

export default {
  name: 'ApiSelect',
  components: { ApiWrapper, ApiWrapperAppend },
  mixins: [apiPropsMixin],
  props: {
    // disable delay per default
    autoSaveDelay: { type: Number, default: 0, required: false }
  },
  data () {
    return {
    }
  }
}
</script>

<style scoped>
</style>
