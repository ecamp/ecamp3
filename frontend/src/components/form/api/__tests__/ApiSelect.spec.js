// Libraries
import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ApiSelect from '../ApiSelect.vue'
import flushPromises from 'flush-promises'
import ApiWrapper from '@/components/form/api/ApiWrapper.vue'
import { i18n } from '@/plugins'
import merge from 'lodash/merge'
import { ApiMock } from '@/components/form/api/__tests__/ApiMock'
import { waitForDebounce } from '@/test/util'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

describe('An ApiSelect', () => {
  let vuetify
  let wrapper
  let apiMock

  const fieldName = 'test-field/123'

  const FIRST_OPTION = {
    value: 1,
    text: 'firstOption',
  }
  const SECOND_OPTION = {
    value: '2',
    text: 'secondOption',
  }

  const selectValues = [FIRST_OPTION, SECOND_OPTION]

  beforeEach(() => {
    vuetify = new Vuetify()
    apiMock = ApiMock.create()
  })

  afterEach(() => {
    jest.restoreAllMocks()
    wrapper.destroy()
  })

  const mount = (options) => {
    const app = Vue.component('App', {
      components: { ApiSelect },
      props: {
        fieldName: { type: String, default: fieldName },
        selectValues: { type: Array, default: () => selectValues },
      },
      template: `
        <div data-app>
        <api-select
          :auto-save="false"
          :fieldname="fieldName"
          uri="test-field/123"
          label="Test field"
          required="true"
          :items="selectValues"
        />
        </div>
      `,
    })
    apiMock.get().thenReturn(ApiMock.success(FIRST_OPTION.value).forFieldName(fieldName))
    const defaultOptions = {
      mocks: {
        $tc: () => {},
        api: apiMock.getMocks(),
      },
    }
    return mountComponent(app, {
      vuetify,
      i18n,
      attachTo: document.body,
      ...merge(defaultOptions, options),
    })
  }

  test('triggers api.patch and status update if input changes', async () => {
    apiMock.patch().thenReturn(ApiMock.success(SECOND_OPTION.value))
    wrapper = mount()

    await flushPromises()

    await wrapper.find('.v-input__slot').trigger('click')
    await wrapper.findAll('[role="option"]').at(1).trigger('click')
    await wrapper.find('input').trigger('submit')

    await waitForDebounce()
    await flushPromises()

    expect(apiMock.getMocks().patch).toBeCalledTimes(1)
    expect(wrapper.findComponent(ApiWrapper).vm.localValue).toBe(SECOND_OPTION.value)
  })

  test('updates state if value in store is refreshed and has new value', async () => {
    wrapper = mount()
    apiMock.get().thenReturn(ApiMock.success(SECOND_OPTION.value).forFieldName(fieldName))

    wrapper.findComponent(ApiWrapper).vm.reload()

    await waitForDebounce()
    await flushPromises()

    expect(wrapper.findComponent(ApiWrapper).vm.localValue).toBe(SECOND_OPTION.value)
    expect(wrapper.html()).toContain(SECOND_OPTION.text)
    expect(wrapper.html()).not.toContain(FIRST_OPTION.text)
  })
})
