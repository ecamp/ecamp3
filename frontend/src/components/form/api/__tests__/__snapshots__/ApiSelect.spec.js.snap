// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiTextField.vue renders correctly 1`] = `
<form
  class="v-form d-flex flex-wrap my-4"
  novalidate="novalidate"
>
  <div
    class="v-input v-input--hide-details theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select"
  >
    <div
      class="v-input__control"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="list-9"
        class="v-input__slot"
        role="button"
      >
        <fieldset
          aria-hidden="true"
        >
          <legend
            style="width: 0px;"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
        <div
          class="v-select__slot"
        >
          <label
            class="v-label theme--light"
            for="input-9"
            style="left: 0px; position: absolute;"
          >
            Test Field
          </label>
          <div
            class="v-select__selections"
          >
            <input
              aria-readonly="false"
              autocomplete="off"
              id="input-9"
              readonly="readonly"
              type="text"
            />
          </div>
          <div
            class="v-input__append-inner"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-menu-down theme--light"
            />
             
            <div
              class="d-flex"
            >
              <!---->
               
              <!---->
               
              <!---->
            </div>
          </div>
          <input
            type="hidden"
            value="Test Value"
          />
        </div>
        <div
          class="v-menu"
        >
          <!---->
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</form>
`;
